invenio_accounts/sessions.py

```diff
@@ -0,0 +1,91 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Invenio; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Backend session management for accounts.
+
+The actual backend sessions are provided by ``flask-kvsession``.
+This module provides management functionality and populates the
+:class:`invenio_accounts.models.SessionActivity` table as new sessions are
+created.
+"""
+
+from __future__ import absolute_import, print_function
+
+import flask
+from flask_login import user_logged_in
+from sqlalchemy.exc import IntegrityError
+from werkzeug.local import LocalProxy
+
+from invenio_accounts.models import SessionActivity
+
+_datastore = LocalProxy(lambda: flask.current_app.
+                        extensions['invenio-accounts'].datastore)
+
+_sessionstore = LocalProxy(lambda: flask.current_app.
+                           extensions['invenio-accounts'].sessionstore)
+
+
+def add_session(session=None):
+    r"""Add a session to the SessionActivity table.
+
+    :param session: Flask Session object to add. If None, ``flask.session``
+        is used. The object is expected to have a dictionary entry named
+        ``"user_id"`` and a field ``sid_s``
+    """
+    user_id, sid_s = session['user_id'], session.sid_s
+    session_activity = SessionActivity(user_id=user_id, sid_s=sid_s)
+    try:
+        _datastore.db.session.add(session_activity)
+        _datastore.commit()
+    except IntegrityError:
+        _datastore.db.session.rollback()
+
+
+def login_listener(app, user):
+    """Connect to the user_logged_in signal for table population."""
+    @flask.after_this_request
+    def populate_session_activity(response):
+        """Add the current session to the SessionActivity table."""
+        # Have to save the session so that the sid_s gets generated.
+        app.session_interface.save_session(app, flask.session, response)
+        add_session(flask.session)
+        return response
+
+
+def delete_session(sid_s):
+    """Delete entries in the data- and kvsessionstore with the given sid_s.
+
+    On a successful deletion, the flask-kvsession store returns 1 while the
+    sqlalchemy datastore returns None.
+
+    :returns: 1 if deletion was successful.
+    """
+    # Remove entries from sessionstore
+    _sessionstore.delete(sid_s)
+    # Find and remove the corresponding SessionActivity entry
+    sessionactivity = _datastore.db.session.query(SessionActivity).\
+        filter_by(sid_s=sid_s).first()
+    _datastore.db.session.delete(sessionactivity)
+    _datastore.commit()
+    return 1
```

Comment:
`_sessionstore.delete` returns 1 if deletion was successful,
`_datastore.db.session.delete` returns nothing if deletion was succesful.

Which I guess was why I had this return 1. Though might it just as well not return anything, seeing as how an error will be thrown if either deletion fails.